File: fabric-audit-agent/core/diagnosis.js
import { detectAll } from './detectors/index.js';
import { createStubReasoner } from '../adapters/reasoner.stub.js';
import { buildHealthScore } from './health-score.js';
import { buildRoadmap } from './roadmap.js';
import { buildCapacityVerdict } from './verdict.js';

/**
 * Run the full offline diagnosis over a facts object: detectors -> stub reasoner
 * -> health score, capacity verdict, prioritized roadmap. No network, no API key.
 * @param {object} facts
 */
export async function diagnose(facts) {
  const flags = detectAll(facts);
  const findings = await createStubReasoner().reason(facts, flags);
  return {
    flags,
    findings,
    health: buildHealthScore(findings),
    verdict: buildCapacityVerdict(facts, flags),
    roadmap: buildRoadmap(findings),
  };
}

/**
 * Render a diagnosis result as a human-readable block of text.
 * @param {{findings:any[], health:any, verdict:any, roadmap:any[]}} result
 * @returns {string}
 */
export function formatDiagnosis({ findings, health, verdict, roadmap }) {
  const L = ['', '================  YOUR ESTATE — DIAGNOSIS  ================', ''];
  if (!findings.length) {
    L.push('No issues detected from the data provided.');
    L.push('(If you expected findings, re-check the values / column mapping above.)');
  } else {
    for (const f of findings) {
      L.push(`[${f.score.level}] ${f.what}`);
      L.push(`    Why:    ${f.why}`);
      L.push(`    Impact: ${f.impact}`);
      L.push(`    Fix:    ${(f.fix ?? [])[0] ?? ''}`);
      L.push('');
    }
    L.push('-----------------------------------------------------------');
    L.push(`Findings: ${findings.length}   Health: ${health.overall}/100`);
    L.push(`By domain: ${JSON.stringify(health.byDomain)}`);
    L.push(`Capacity verdict: ${verdict.decision.toUpperCase()} — ${verdict.reason}`);
    L.push('', 'Do these first:');
    roadmap.slice(0, 5).forEach(r => L.push(`   #${r.rank} [${r.level}] ${r.what}`));
  }
  L.push('', '===========================================================', '');
  return L.join('\n');
}

File: fabric-audit-agent-py/fabric_audit_agent/diagnosis.py
"""Full offline diagnosis. Faithful port of the Node ``core/diagnosis.js``.

detectors -> stub reasoner -> health score, capacity verdict, prioritized roadmap.
No network, no API key.
"""
import json
from .detectors import detect_all
from .reasoner_stub import create_stub_reasoner
from .health_score import build_health_score
from .roadmap import build_roadmap
from .verdict import build_capacity_verdict


def diagnose(facts):
    flags = detect_all(facts)
    findings = create_stub_reasoner()["reason"](facts, flags)
    return {
        "flags": flags,
        "findings": findings,
        "health": build_health_score(findings),
        "verdict": build_capacity_verdict(facts, flags),
        "roadmap": build_roadmap(findings),
    }


def format_diagnosis(result):
    findings, health = result["findings"], result["health"]
    verdict, roadmap = result["verdict"], result["roadmap"]
    lines = ["", "================  YOUR ESTATE — DIAGNOSIS  ================", ""]
    if not findings:
        lines.append("No issues detected from the data provided.")
        lines.append("(If you expected findings, re-check the values / column mapping above.)")
    else:
        for f in findings:
            fix0 = (f.get("fix") or [None])[0]
            lines.append(f"[{f['score']['level']}] {f['what']}")
            lines.append(f"    Why:    {f['why']}")
            lines.append(f"    Impact: {f['impact']}")
            lines.append(f"    Fix:    {fix0 if fix0 is not None else ''}")
            lines.append("")
        lines.append("-----------------------------------------------------------")
        lines.append(f"Findings: {len(findings)}   Health: {health['overall']}/100")
        lines.append(f"By domain: {json.dumps(health['byDomain'])}")
        lines.append(f"Capacity verdict: {verdict['decision'].upper()} — {verdict['reason']}")
        lines.append("")
        lines.append("Do these first:")
        for r in roadmap[:5]:
            lines.append(f"   #{r['rank']} [{r['level']}] {r['what']}")
    lines += ["", "===========================================================", ""]
    return "\n".join(lines)
